README.md
# skeleton_text
  
A package provides an easy way to add skeleton text loading animation in Flutter project  
  
<p>
  <img src="https://raw.githubusercontent.com/101Loop/Skeleton-Text/master/assets/preview2.gif" height="50%" width="50%" />
</p>
 
## Dependency 
```
dependencies:
  skeleton_text: 
```
 
## How To Use  
  
  ```
  import 'package:skeleton_text/skeleton_text.dart';
  ```

> To achieve the above example animation use the following code :

     ListView.builder(
       scrollDirection: Axis.vertical,
       physics: BouncingScrollPhysics(),
       itemCount: 5,
       itemBuilder: (BuildContext context, int index) {
         return Container(
           height: 240,
           margin: EdgeInsets.symmetric(horizontal: 20),
           child: Row(
             children: [
               Expanded(
                 child: SkeletonAnimation(
                   shimmerColor: Colors.grey,
                   borderRadius: BorderRadius.circular(20),
                   shimmerDuration: 1000,
                   child: Container(
                     decoration: BoxDecoration(
                       color: Colors.grey[300],
                       borderRadius: BorderRadius.circular(20),
                       boxShadow: shadowList,
                     ),
                     margin: EdgeInsets.only(top: 40),
                   ),
                 ),
               ),
               Expanded(
                 child: Container(
                   margin: EdgeInsets.only(top: 60, bottom: 20),
                   decoration: BoxDecoration(
                     color: Colors.grey,
                     boxShadow: shadowList,
                     borderRadius: BorderRadius.only(
                       topRight: Radius.circular(20),
                       bottomRight: Radius.circular(20),
                     ),
                   ),
                   child: Column(
                     mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                     crossAxisAlignment: CrossAxisAlignment.start,
                     mainAxisSize: MainAxisSize.max,
                     children: <Widget>[
                       Padding(
                         padding: const EdgeInsets.only(left: 15.0, bottom: 5.0),
                         child: SkeletonAnimation(
                           borderRadius: BorderRadius.circular(10.0),
                           shimmerColor: index % 2 != 0 ? Colors.grey : Colors.white54,
                           child: Container(
                             height: 30,
                             width: MediaQuery.of(context).size.width * 0.35,
                             decoration: BoxDecoration(
                                 borderRadius: BorderRadius.circular(10.0),
                                 color: Colors.grey[300]),
                           ),
                         ),
                       ),
                       Padding(
                         padding: const EdgeInsets.only(left: 15.0),
                         child: Padding(
                           padding: const EdgeInsets.only(right: 5.0),
                           child: SkeletonAnimation(
                             borderRadius: BorderRadius.circular(10.0),
                             shimmerColor: index % 2 != 0 ? Colors.grey : Colors.white54,
                             child: Container(
                               width: 60,
                               height: 30,
                               decoration: BoxDecoration(
                                   borderRadius: BorderRadius.circular(10.0),
                                   color: Colors.grey[300]),
                             ),
                           ),
                         ),
                       ),
                     ],
                   ),
                 ),
               ),
             ],
           ),
         );
       },
     )

## Explanation 

I have created a List builder of 5 Containers. Wrapping any widget by SkeletonAnimation would give a beautiful loading animation.

## SkeletonAnimation<T> Constructor :

```
  SkeletonAnimation({
      @required this.child,
      this.shimmerColor = Colors.white54,
      this.gradientColor = const Color.fromARGB(0, 244, 244, 244),
      this.curve = Curves.fastOutSlowIn,
      this.borderRadius = const BorderRadius.all(Radius.circular(0)),
      this.shimmerDuration = 1000,
      Key key,
    })
```


CHANGELOG.md
## [3.0.1] - 2023-08-22
* Fix issue where BorderRadius? can't be nullable

## [3.0.0] - 2021-03-03
* Upgraded to Flutter 2.0 with full sound null-safety support

## [2.0.2] - 2021-01-22
* Added preview GIF.

## [2.0.1] - 2021-01-22
* Fixed README

## [2.0.0] - 2021-01-22
* Added shimmer clipping feature.
* Added new example.
* Shimmer animation speed is now customizable.

## [1.0.0] - 2019-04-19
* Added Example Project.
* The package is now stable enough.

## [0.0.2] - 2019-04-19
* Added Documentation.

## [0.0.1] - 2019-04-19
* Initial release.


test/skeleton_text_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:skeleton_text/skeleton_text.dart';

void main() {
  test('Adds skeleton text animation to a Widget', () {
    final skeletonText = SkeletonAnimation(
      child: Container(
        decoration: BoxDecoration(
          borderRadius: new BorderRadius.circular(30.0),
          color: Colors.grey[300],
        ),
        width: 30.0,
        height: 13.0,
      ),
    );

    skeletonText.createElement();
  });
}


lib/skeleton_text.dart
///
/// author: Mahen Gandhi <mahengandhi19@gmail.com>
/// github: https://github.com/imlegend19
///

library skeleton_text;

import 'package:flutter/material.dart';

class SkeletonAnimation extends StatefulWidget {
  final Widget child;
  final Color shimmerColor;
  final Color gradientColor;
  final Curve curve;
  final BorderRadiusGeometry borderRadius;
  final int shimmerDuration;

  SkeletonAnimation({
    required this.child,
    this.shimmerColor = Colors.white54,
    this.gradientColor = const Color.fromARGB(0, 244, 244, 244),
    this.curve = Curves.fastOutSlowIn,
    this.borderRadius = const BorderRadius.all(Radius.circular(0)),
    this.shimmerDuration = 1000,
    Key? key,
  }) : super(key: key);

  @override
  _SkeletonAnimationState createState() => _SkeletonAnimationState();
}

class _SkeletonAnimationState extends State<SkeletonAnimation>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void initState() {
    super.initState();

    _controller = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: widget.shimmerDuration),
    );

    _controller.repeat();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  E? firstOrNull<E>(Iterable<DiagnosticsNode> list) =>
      list.isEmpty ? null : list.first.value as E?;

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: <Widget>[
        widget.child,
        Positioned.fill(
          child: Container(
            color: Colors.transparent,
            margin: firstOrNull(widget.child
                .toDiagnosticsNode()
                .getProperties()
                .where((element) => element.name == "margin")),
            child: ClipRRect(
              borderRadius: widget.borderRadius,
              child: AnimatedBuilder(
                animation: _controller,
                builder: (BuildContext context, Widget? child) {
                  return FractionallySizedBox(
                    widthFactor: .2,
                    alignment: AlignmentGeometryTween(
                      begin: Alignment(-1.0 - .2 * 3, .0),
                      end: Alignment(1.0 + .2 * 3, .0),
                    )
                        .chain(CurveTween(curve: widget.curve))
                        .evaluate(_controller)!,
                    child: child,
                  );
                },
                child: DecoratedBox(
                  decoration: BoxDecoration(
                    gradient: LinearGradient(
                      colors: [
                        widget.gradientColor,
                        widget.shimmerColor,
                      ],
                    ),
                  ),
                ),
              ),
            ),
          ),
        )
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:skeleton_text/skeleton_text.dart';

void main() => runApp(const MyApp());

List<BoxShadow> shadowList = [
  BoxShadow(
      color: Colors.grey[300]!, blurRadius: 30, offset: const Offset(0, 10))
];

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Skeleton Text Demo",
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        backgroundColor: Colors.black54,
        body: ListView.builder(
          scrollDirection: Axis.vertical,
          physics: const BouncingScrollPhysics(),
          itemCount: 5,
          itemBuilder: (BuildContext context, int index) {
            return Container(
              height: 240,
              margin: const EdgeInsets.symmetric(horizontal: 20),
              child: Row(
                children: [
                  Expanded(
                    child: SkeletonAnimation(
                      shimmerColor: Colors.grey,
                      borderRadius: BorderRadius.circular(20),
                      shimmerDuration: 1000,
                      child: Container(
                        decoration: BoxDecoration(
                          color: Colors.grey[300],
                          borderRadius: BorderRadius.circular(20),
                          boxShadow: shadowList,
                        ),
                        margin: const EdgeInsets.only(top: 40),
                      ),
                    ),
                  ),
                  Expanded(
                    child: Container(
                      margin: const EdgeInsets.only(top: 60, bottom: 20),
                      decoration: BoxDecoration(
                        color: Colors.grey,
                        boxShadow: shadowList,
                        borderRadius: const BorderRadius.only(
                          topRight: Radius.circular(20),
                          bottomRight: Radius.circular(20),
                        ),
                      ),
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                        crossAxisAlignment: CrossAxisAlignment.start,
                        mainAxisSize: MainAxisSize.max,
                        children: <Widget>[
                          Padding(
                            padding:
                                const EdgeInsets.only(left: 15.0, bottom: 5.0),
                            child: SkeletonAnimation(
                              borderRadius: BorderRadius.circular(10.0),
                              shimmerColor:
                                  index % 2 != 0 ? Colors.grey : Colors.white54,
                              child: Container(
                                height: 30,
                                width: MediaQuery.of(context).size.width * 0.35,
                                decoration: BoxDecoration(
                                    borderRadius: BorderRadius.circular(10.0),
                                    color: Colors.grey[300]),
                              ),
                            ),
                          ),
                          Padding(
                            padding: const EdgeInsets.only(left: 15.0),
                            child: Padding(
                              padding: const EdgeInsets.only(right: 5.0),
                              child: SkeletonAnimation(
                                borderRadius: BorderRadius.circular(10.0),
                                shimmerColor: index % 2 != 0
                                    ? Colors.grey
                                    : Colors.white54,
                                child: Container(
                                  width: 60,
                                  height: 30,
                                  decoration: BoxDecoration(
                                      borderRadius: BorderRadius.circular(10.0),
                                      color: Colors.grey[300]),
                                ),
                              ),
                            ),
                          ),
                        ],
                      ),
                    ),
                  ),
                ],
              ),
            );
          },
        ),
      ),
    );
  }
}
